routing/transit_graph.cpp

```diff
@@ -0,0 +1,230 @@
+}
+
+bool TransitGraph::FindReal(Segment const & fake, Segment & real) const
+{
+  return m_fake.FindReal(fake, real);
+}
+
+void TransitGraph::Fill(vector<transit::Stop> const & stops, vector<transit::Gate> const & gates,
+                        vector<transit::Edge> const & edges, EdgeEstimator const & estimator,
+                        NumMwmId numMwmId, IndexGraph & indexGraph)
+{
+  m_mwmId = numMwmId;
+  map<transit::StopId, Junction> stopCoords;
+  for (auto const & stop : stops)
+    stopCoords[stop.GetId()] = Junction(stop.GetPoint(), feature::kDefaultAltitudeMeters);
+
+  for (auto const & gate : gates)
+  {
+    // TODO (@t.yan) after https://github.com/mapsme/omim/pull/7240 merge
+    // auto const gateSegment = gate.GetBestPedestrianSegment();
+    // Segment real(numMwmId, gateSegment.GetFeatureId(), gateSegment.GetSegmentIdx(), gateSegment.GetForward());
+    auto const ending = MakeFakeEnding( Segment() /* real */, gate.GetPoint(), estimator, indexGraph);
+    if (gate.GetEntrance())
+      AddGate(gate, ending, stopCoords, true /* isEnter */);
+    if (gate.GetExit())
+      AddGate(gate, ending, stopCoords, false /* isEnter */);
+  }
+
+  map<transit::StopId, set<transit::Edge>> outgoing;
+  map<transit::StopId, set<transit::Edge>> ingoing;
+  for (auto const & edge : edges)
+  {
+    AddEdge(edge, stopCoords);
+    outgoing[edge.GetStartStopId()].insert(edge);
+    ingoing[edge.GetFinishStopId()].insert(edge);
+  }
+
+  AddConnections(outgoing, true /* isOutgoing */);
+  AddConnections(ingoing, false /* isOutgoing */);
+}
+
+Segment TransitGraph::GetTransitSegment(uint32_t segmentIdx) const
+{
+  return Segment(m_mwmId, kTransitFeatureId, segmentIdx, false);
+}
+
+Segment TransitGraph::GetNewTransitSegment() const
+{
+  CHECK_LESS_OR_EQUAL(m_fake.GetSize(), std::numeric_limits<uint32_t>::max(), ());
+  return GetTransitSegment(static_cast<uint32_t>(m_fake.GetSize()) /* segmentIdx */);
+}
+
+void TransitGraph::AddGate(transit::Gate const & gate, FakeEnding const & ending,
+                           map<transit::StopId, Junction> const & stopCoords, bool isEnter)
+{
+  Segment const dummy = Segment();
+  for (auto const & projection : ending.m_projections)
+  {
+    // Add projection edges
+    auto const projectionSegment = GetNewTransitSegment();
+    FakeVertex projectionVertex(isEnter ? projection.m_junction : ending.m_originJunction,
+                                isEnter ? ending.m_originJunction : projection.m_junction,
+                                FakeVertex::Type::PureFake);
+    m_fake.AddStandaloneVertex(projectionSegment, projectionVertex);
+
+    // Add fake parts of real
+    FakeVertex forwardPartOfReal(isEnter ? projection.m_segmentBack : projection.m_junction,
+                                 isEnter ? projection.m_junction : projection.m_segmentFront,
+                                 FakeVertex::Type::PartOfReal);
+    auto const fakeForwardSegment = GetNewTransitSegment();
```

Comment:
Actually, what's the real difference between these transit segments and fake segments?

Reply:
I mean, looks like transit segments and fake segments are nonexistent, therefore do we really need to handle them separately?

Reply:
Ah, see now, these segments are added to m_fake, nice.